### PR TITLE
🐛 Stopping container without starting them no longer raises error

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

If there is no compose-spec stored in the dy-sidecar, when it's asked to stop the containers:
- **Previously:** it currently raised an error
- **Now:** logs a warning

When did this happen? When the study was opened and closed as soon as the service went to pending. Because it did not have time to properly start, it only executed the shutdown sequence.

## Related issue/s

- https://github.com/ITISFoundation/osparc-simcore/issues/3581
- fixes https://github.com/ITISFoundation/osparc-simcore/issues/3578
- https://github.com/ITISFoundation/osparc-issues/issues/638
## Checklist

- [x] Unit tests for the changes exist

### CODE REVIEW

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/long_running_tasks.py [hunk 1]
```diff
@@ -141,7 +141,9 @@ async def task_runs_docker_compose_down(
     settings: ApplicationSettings,
 ) -> None:
     if shared_store.compose_spec is None:
-        raise RuntimeError("No compose-spec was found")
+        logger.warning("No compose-spec was found")
+        progress.update(message="done", percent=0.99)
```
Comment (sanderegg) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/long_running_tasks.py:145]: Q: progress 0.99??? is this not already automatically set to 1 and completed?

Comment (sanderegg) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/long_running_tasks.py:145]: Q: progress 0.99??? is this not already automatically set to 1 and completed?

File: services/dynamic-sidecar/tests/unit/test_api_containers_long_running_tasks.py [hunk 2]
```diff
@@ -446,17 +446,18 @@ async def test_containers_down_after_starting(
...
-            assert result is None
-    assert 'RuntimeError("No compose-spec was found")' in f"{exec_info.value}"
+    async with periodic_task_result(
+        client=client,
+        task_id=await _get_task_id_docker_compose_down(httpx_async_client),
+        task_timeout=CREATE_SERVICE_CONTAINERS_TIMEOUT,
+        status_poll_interval=FAST_STATUS_POLL,
+    ) as result:
+        assert result is None
+    assert "No compose-spec was found" in caplog_info_debug.text
```
Comment (sanderegg) [inline services/dynamic-sidecar/tests/unit/test_api_containers_long_running_tasks.py:460]: I am not sure we need to check the logs unless we write something very specific like an error code. the test already checks that it does not raise which is enough right?

**Review by mrnicegyu11 — COMMENTED** [review]
Very nice, could you please give me a quick summary of why this corner case happens? While this fix you provided will surely make the platform more stable by not raising exceptions & stopping the dy-sidecar, it does not address the core of the issue I guess. This is, if I get this correctly, that a dy-sidecar expects to find a docker-compose spec (yml file?) in a shared_store (what is this actually, some place that keeps a state, shared by all sidecars?), but it is not there. So I wonder in which order the assoc. microservices talk to each other and at which step in the pipeline the docker-compose spec goes missing :--) I would love if you could give me some bullet points on this.

**Review by sanderegg — APPROVED** [review]
great! thanks for taking care of this!

**Comment by mrnicegyu11** [comment]
I just read the assoc. issue and I guess you guys have already fixed the issue of the director-v2 not providing a docker-compose spec? sorry for the stupid questions :)

**Comment by GitHK** [comment]
Let's see if I can answer your questions

> Very nice, could you please give me a quick summary of why this corner case happens? 

If you open a study with a service, and close it as soon as you see the service's state going to `pending` corner case will happen.
Since there was not enough time to start, no containers were created. When you have no containers created, no compose-spec is stored in the shared-store.

> While this fix you provided will surely make the platform more stable by not raising exceptions & stopping the dy-sidecar, it does not address the core of the issue I guess.

I think does address the issue at hand, as per the PR's title. If you are referring to https://github.com/ITISFoundation/osparc-simcore/issues/3581 then this is just an improvement.

> This is, if I get this correctly, that a dy-sidecar expects to find a docker-compose spec (yml file?) in a shared_store (what is this actually, some place that keeps a state, shared by all sidecars?), but it is not there. 

The name `shared_store` is just an area of memory where the dy-sidecar stores information ti will require in the future. One of these is the `compose-spec`. This information is also persisted to disk and stored in the docker-volumes, to allow for reboot of the dy-sidecar.

> So I wonder in which order the assoc. microservices talk to each other and at which step in the pipeline the docker-compose spec goes missing :--) I would love if you could give me some bullet points on this.

Let's see if this answer is enough:
- view the dy-sidecar as a slave which receives commands and executes them, as such it is the last in a chain.
- the director-v2 is the interface to the dy-sidecar, it sends commands
- In this situation the director-v2 sent a shutdown command a bit too early than when it was expected (our corner case)

**Review by mrnicegyu11 — APPROVED** [review]
Thanks a lot! super!

**Review by pcrespov — APPROVED** [review]
Q: But if it has no compose-specs .. it can basically do nothing. Just waiting the instruction of the director to shut down, right? 

This is a flaw that, as you know, could be overcome if the specs are passed directly with the sidecar as it starts ... an dnot transmitted via the API :-)
